File: src/frontend/tributacao/lib/app/data/model/tribut_icms_uf_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:tributacao/app/data/domain/domain_imports.dart';

class TributIcmsUfModel {
	int? id;
	int? idTributConfiguraOfGt;
	String? ufDestino;
	String? cst;
	String? csosn;
	String? modalidadeBc;
	int? cfop;
	double? aliquota;
	double? valorPauta;
	double? valorPrecoMaximo;
	double? mva;
	double? porcentoBc;
	String? modalidadeBcSt;
	double? aliquotaInternaSt;
	double? aliquotaInterestadualSt;
	double? porcentoBcSt;
	double? aliquotaIcmsSt;
	double? valorPautaSt;
	double? valorPrecoMaximoSt;

	TributIcmsUfModel({
		this.id,
		this.idTributConfiguraOfGt,
		this.ufDestino,
		this.cst,
		this.csosn,
		this.modalidadeBc,
		this.cfop,
		this.aliquota,
		this.valorPauta,
		this.valorPrecoMaximo,
		this.mva,
		this.porcentoBc,
		this.modalidadeBcSt,
		this.aliquotaInternaSt,
		this.aliquotaInterestadualSt,
		this.porcentoBcSt,
		this.aliquotaIcmsSt,
		this.valorPautaSt,
		this.valorPrecoMaximoSt,
	});

	static List<String> dbColumns = <String>[
		'id',
		'uf_destino',
		'cst',
		'csosn',
		'modalidade_bc',
		'cfop',
		'aliquota',
		'valor_pauta',
		'valor_preco_maximo',
		'mva',
		'porcento_bc',
		'modalidade_bc_st',
		'aliquota_interna_st',
		'aliquota_interestadual_st',
		'porcento_bc_st',
		'aliquota_icms_st',
		'valor_pauta_st',
		'valor_preco_maximo_st',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Uf Destino',
		'Cst',
		'Csosn',
		'Modalidade Bc',
		'Cfop',
		'Aliquota',
		'Valor Pauta',
		'Valor Preco Maximo',
		'Mva',
		'Porcento Bc',
		'Modalidade Bc St',
		'Aliquota Interna St',
		'Aliquota Interestadual St',
		'Porcento Bc St',
		'Aliquota Icms St',
		'Valor Pauta St',
		'Valor Preco Maximo St',
	];

	TributIcmsUfModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idTributConfiguraOfGt = jsonData['idTributConfiguraOfGt'];
		ufDestino = TributIcmsUfDomain.getUfDestino(jsonData['ufDestino']);
		cst = TributIcmsUfDomain.getCst(jsonData['cst']);
		csosn = TributIcmsUfDomain.getCsosn(jsonData['csosn']);
		modalidadeBc = TributIcmsUfDomain.getModalidadeBc(jsonData['modalidadeBc']);
		cfop = jsonData['cfop'];
		aliquota = jsonData['aliquota']?.toDouble();
		valorPauta = jsonData['valorPauta']?.toDouble();
		valorPrecoMaximo = jsonData['valorPrecoMaximo']?.toDouble();
		mva = jsonData['mva']?.toDouble();
		porcentoBc = jsonData['porcentoBc']?.toDouble();
		modalidadeBcSt = TributIcmsUfDomain.getModalidadeBcSt(jsonData['modalidadeBcSt']);
		aliquotaInternaSt = jsonData['aliquotaInternaSt']?.toDouble();
		aliquotaInterestadualSt = jsonData['aliquotaInterestadualSt']?.toDouble();
		porcentoBcSt = jsonData['porcentoBcSt']?.toDouble();
		aliquotaIcmsSt = jsonData['aliquotaIcmsSt']?.toDouble();
		valorPautaSt = jsonData['valorPautaSt']?.toDouble();
		valorPrecoMaximoSt = jsonData['valorPrecoMaximoSt']?.toDouble();
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idTributConfiguraOfGt'] = idTributConfiguraOfGt != 0 ? idTributConfiguraOfGt : null;
		jsonData['ufDestino'] = TributIcmsUfDomain.setUfDestino(ufDestino);
		jsonData['cst'] = TributIcmsUfDomain.setCst(cst);
		jsonData['csosn'] = TributIcmsUfDomain.setCsosn(csosn);
		jsonData['modalidadeBc'] = TributIcmsUfDomain.setModalidadeBc(modalidadeBc);
		jsonData['cfop'] = cfop;
		jsonData['aliquota'] = aliquota;
		jsonData['valorPauta'] = valorPauta;
		jsonData['valorPrecoMaximo'] = valorPrecoMaximo;
		jsonData['mva'] = mva;
		jsonData['porcentoBc'] = porcentoBc;
		jsonData['modalidadeBcSt'] = TributIcmsUfDomain.setModalidadeBcSt(modalidadeBcSt);
		jsonData['aliquotaInternaSt'] = aliquotaInternaSt;
		jsonData['aliquotaInterestadualSt'] = aliquotaInterestadualSt;
		jsonData['porcentoBcSt'] = porcentoBcSt;
		jsonData['aliquotaIcmsSt'] = aliquotaIcmsSt;
		jsonData['valorPautaSt'] = valorPautaSt;
		jsonData['valorPrecoMaximoSt'] = valorPrecoMaximoSt;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idTributConfiguraOfGt = plutoRow.cells['idTributConfiguraOfGt']?.value;
		ufDestino = plutoRow.cells['ufDestino']?.value != '' ? plutoRow.cells['ufDestino']?.value : 'AC';
		cst = plutoRow.cells['cst']?.value != '' ? plutoRow.cells['cst']?.value : '00';
		csosn = plutoRow.cells['csosn']?.value != '' ? plutoRow.cells['csosn']?.value : '101';
		modalidadeBc = plutoRow.cells['modalidadeBc']?.value != '' ? plutoRow.cells['modalidadeBc']?.value : '0-Margem Valor Agregado';
		cfop = plutoRow.cells['cfop']?.value;
		aliquota = plutoRow.cells['aliquota']?.value?.toDouble();
		valorPauta = plutoRow.cells['valorPauta']?.value?.toDouble();
		valorPrecoMaximo = plutoRow.cells['valorPrecoMaximo']?.value?.toDouble();
		mva = plutoRow.cells['mva']?.value?.toDouble();
		porcentoBc = plutoRow.cells['porcentoBc']?.value?.toDouble();
		modalidadeBcSt = plutoRow.cells['modalidadeBcSt']?.value != '' ? plutoRow.cells['modalidadeBcSt']?.value : '0-Valor Preço Máximo';
		aliquotaInternaSt = plutoRow.cells['aliquotaInternaSt']?.value?.toDouble();
		aliquotaInterestadualSt = plutoRow.cells['aliquotaInterestadualSt']?.value?.toDouble();
		porcentoBcSt = plutoRow.cells['porcentoBcSt']?.value?.toDouble();
		aliquotaIcmsSt = plutoRow.cells['aliquotaIcmsSt']?.value?.toDouble();
		valorPautaSt = plutoRow.cells['valorPautaSt']?.value?.toDouble();
		valorPrecoMaximoSt = plutoRow.cells['valorPrecoMaximoSt']?.value?.toDouble();
	}	

	TributIcmsUfModel clone() {
		return TributIcmsUfModel(
			id: id,
			idTributConfiguraOfGt: idTributConfiguraOfGt,
			ufDestino: ufDestino,
			cst: cst,
			csosn: csosn,
			modalidadeBc: modalidadeBc,
			cfop: cfop,
			aliquota: aliquota,
			valorPauta: valorPauta,
			valorPrecoMaximo: valorPrecoMaximo,
			mva: mva,
			porcentoBc: porcentoBc,
			modalidadeBcSt: modalidadeBcSt,
			aliquotaInternaSt: aliquotaInternaSt,
			aliquotaInterestadualSt: aliquotaInterestadualSt,
			porcentoBcSt: porcentoBcSt,
			aliquotaIcmsSt: aliquotaIcmsSt,
			valorPautaSt: valorPautaSt,
			valorPrecoMaximoSt: valorPrecoMaximoSt,
		);			
	}

	
}

File: src/backend/tributacao/src/model/tribut_icms_uf_model.py
from src import db


class TributIcmsUfModel(db.Model):
    __tablename__ = 'tribut_icms_uf'

    id = db.Column(db.Integer, primary_key=True)
    uf_destino = db.Column(db.String(2))
    cst = db.Column(db.String(2))
    csosn = db.Column(db.String(3))
    modalidade_bc = db.Column(db.String(1))
    cfop = db.Column(db.Integer)
    aliquota = db.Column(db.Float)
    valor_pauta = db.Column(db.Float)
    valor_preco_maximo = db.Column(db.Float)
    mva = db.Column(db.Float)
    porcento_bc = db.Column(db.Float)
    modalidade_bc_st = db.Column(db.String(1))
    aliquota_interna_st = db.Column(db.Float)
    aliquota_interestadual_st = db.Column(db.Float)
    porcento_bc_st = db.Column(db.Float)
    aliquota_icms_st = db.Column(db.Float)
    valor_pauta_st = db.Column(db.Float)
    valor_preco_maximo_st = db.Column(db.Float)
    id_tribut_configura_of_gt = db.Column(db.Integer, db.ForeignKey('tribut_configura_of_gt.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_tribut_configura_of_gt = data.get('idTributConfiguraOfGt')
        self.uf_destino = data.get('ufDestino')
        self.cst = data.get('cst')
        self.csosn = data.get('csosn')
        self.modalidade_bc = data.get('modalidadeBc')
        self.cfop = data.get('cfop')
        self.aliquota = data.get('aliquota')
        self.valor_pauta = data.get('valorPauta')
        self.valor_preco_maximo = data.get('valorPrecoMaximo')
        self.mva = data.get('mva')
        self.porcento_bc = data.get('porcentoBc')
        self.modalidade_bc_st = data.get('modalidadeBcSt')
        self.aliquota_interna_st = data.get('aliquotaInternaSt')
        self.aliquota_interestadual_st = data.get('aliquotaInterestadualSt')
        self.porcento_bc_st = data.get('porcentoBcSt')
        self.aliquota_icms_st = data.get('aliquotaIcmsSt')
        self.valor_pauta_st = data.get('valorPautaSt')
        self.valor_preco_maximo_st = data.get('valorPrecoMaximoSt')

    def serialize(self):
        return {
            'id': self.id,
            'idTributConfiguraOfGt': self.id_tribut_configura_of_gt,
            'ufDestino': self.uf_destino,
            'cst': self.cst,
            'csosn': self.csosn,
            'modalidadeBc': self.modalidade_bc,
            'cfop': self.cfop,
            'aliquota': self.aliquota,
            'valorPauta': self.valor_pauta,
            'valorPrecoMaximo': self.valor_preco_maximo,
            'mva': self.mva,
            'porcentoBc': self.porcento_bc,
            'modalidadeBcSt': self.modalidade_bc_st,
            'aliquotaInternaSt': self.aliquota_interna_st,
            'aliquotaInterestadualSt': self.aliquota_interestadual_st,
            'porcentoBcSt': self.porcento_bc_st,
            'aliquotaIcmsSt': self.aliquota_icms_st,
            'valorPautaSt': self.valor_pauta_st,
            'valorPrecoMaximoSt': self.valor_preco_maximo_st,
        }